File: ext/llruby/iseq_compiler.cc
#include "llruby/iseq_compiler.h"
#include "llvm/IR/IRBuilder.h"
#include "llvm/IR/Module.h"
#include "llvm/IR/Function.h"
#include "llvm/IR/LLVMContext.h"

extern llvm::LLVMContext context;

namespace llruby {

// IseqCompiler#compile_internal
// @param  [Array] ruby_iseq - Return value of RubyVM::InstructionSequence#to_a
// @return [LLRuby::LLVMFunction]
llvm::Function* IseqCompiler::Compile(const Iseq& iseq) {
  std::vector<llvm::Type*> args = { llvm::IntegerType::get(context, 64) };
  llvm::FunctionType *func_type = llvm::FunctionType::get(llvm::Type::getInt64Ty(context), args, false);
  llvm::Function *func = llvm::Function::Create(func_type, llvm::Function::ExternalLinkage, "rb_hello_func");

  llvm::IRBuilder<> builder(context);
  builder.SetInsertPoint(llvm::BasicBlock::Create(context, "", func));
  builder.CreateRet(builder.getInt64(Qnil));
  return func;
}

} // namespace llruby

File: lib/llruby/iseq_compiler.rb
require 'llruby/llruby'

module LLRuby
  module IseqCompiler
    # @param  [RubyVM::InstructionSequence] iseq
    # @return [LLRuby::LLVMIR]
    def self.compile(iseq)
      unless iseq.is_a?(RubyVM::InstructionSequence)
        raise ArgumentError.new("wrong argument type #{iseq.class} is given! (expected RubyVM::InstructionSequence)")
      end
      compile_internal(iseq.to_a)
    end

    # defined in ext/llruby/llruby.cc
    private_class_method :compile_internal
  end
end
